velox/type/StringToFloatParser.cpp

```diff
@@ -0,0 +1,202 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/type/StringToFloatParser.h"
+
+namespace facebook::velox::util {
+
+static inline bool characterIsSpace(char c) {
+  return c == ' ' || c == '\t' || c == '\n' || c == '\v' || c == '\f' ||
+      c == '\r';
+}
+
+static inline bool
+isEither(const char* str, int offset, char lower, char upper) {
+  char ch = *(str + offset);
+  return ch == lower || ch == upper;
+}
+
+static inline bool isInfinityString(const char* str, size_t length) {
+  // length = 3: inf/INF
+  // length = 8: infinity
+  if (length != 3 && length != 8) {
+    return false;
+  }
+
+  if (isEither(str, 0, 'i', 'I')) {
+    if (isEither(str, 1, 'n', 'N')) {
+      if (isEither(str, 2, 'f', 'F')) {
+        if (length == 3) {
+          return true;
+        }
+
+        if (isEither(str, 3, 'i', 'I')) {
+          if (isEither(str, 4, 'n', 'N')) {
+            if (isEither(str, 5, 'i', 'I')) {
+              if (isEither(str, 6, 't', 'T')) {
+                if (isEither(str, 7, 'y', 'Y')) {
+                  return true;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+static inline bool isNanString(const char* str, size_t length) {
+  if (length != 3) {
+    return false;
+  }
+
+  if (isEither(str, 0, 'n', 'N')) {
+    if (isEither(str, 1, 'a', 'A')) {
+      if (isEither(str, 2, 'n', 'N')) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+template <typename T>
+Status StringToFloatParser::parse(const std::string_view& str, T& out) {
+  auto length = str.length();
+  if (UNLIKELY(length <= 0)) {
+    return Status::Invalid("empty string");
+  }
+  int i = 0;
+  // skip any spaces at the start
+  for (; i < length; ++i) {
+    if (!characterIsSpace(str[i])) {
+      break;
+    }
+  }
+
+  // skip any spaces at the end
+  int j = length - 1;
+  for (; j >= i; j--) {
+    if (!characterIsSpace(str[j])) {
+      break;
+    }
+  }
+
+  if (UNLIKELY(i > j)) {
+    return Status::Invalid("i > j");
+  }
+
+  bool negative = false;
+  // skip leading +/-
+  switch (str[i]) {
+    case '-':
+      negative = true;
+      i++;
+      break;
+    case '+':
+      i++;
+  }
+
+  if (UNLIKELY(i > j)) {
+    return Status::Invalid("i > j");
+  }
+
+  // check INF/INFINITY/NAN, ignoring case
+  for (int k = i; k <= j; ++k) {
+    if (str[k] == 'i' || str[k] == 'I') {
```

Comment:
Why not call isInfinityString directly without checking first character?

Reply:
Yes, I'll refactor it.